### PR TITLE
Zarr prefetching

### LINKED ISSUE #126: Prefetch and cache slices
## Use Case

Playback in XY mode is too slow if we are fetching single slices.

## Solution

We can optimize by prefetching and caching slices.  When time or z slider is moved, prefetch some range of slices (like the next 20, for example.  Or possibly previous 5 / next 15 etc).  Cancel any previous prefetching that is out of range.  XY rendering can then get its data from the local cache.

Choose a simple heuristic to get things working. We can analyze / profile / refine / improve in other tickets once any kind of prefetching is working.

### PR DESCRIPTION
**Pull request by frasercl**
Adds configurable prefetching behavior to `OmeZarrLoader`. Should (at last!) resolve #126. I think I have this feature working pretty well now, but please pull it and try for yourself!

### How it works for users:

- When any zarr volume with a time series loads, the loader begins prefetching adjacent time steps. Moving to adjacent time steps gets faster because those steps may already have been requested and received.
- When the user switches to z-slice mode and the chunking of the volume affords loading a subset of the volume at higher resolution, the loader begins prefetching adjacent chunks in z. Moving to adjacent z-slices gets faster too.
- If both of the above are true, it does both at the same time!

### How it works for developers:

- `RequestQueue` now has two "lanes:" regular and low-priority.
  - The zarr loader uses the regular lane for normal loads and the low-priority lane for prefetches.
  - All regular requests must be started before any low-priority requests get to go.
  - The low-priority lane has its own concurrency limit, which may be set lower (but cannot be set higher!) than the limit for the regular lane. This allows the user to ensure that the queue always leaves some space open for new regular requests to begin immediately.
- `OmeZarrLoader` now properly uses `SubscribableRequestQueue`, which has new method arguments to use `RequestQueue`'s new priority features.
- `OmeZarrLoader` has two new options to configure prefetching behavior:
  - `maxPrefetchDistance` is a `Vector4` representing the maximum number of chunks to prefetch outwards in both directions of a given dimension. (X and Y are of course not currently used in practice, but they're still in there for just in case.) For instance, if a load requests chunks with z coords 3 and 4 and `maxPrefetchDistance.z` is 2, the loader will prefetch similar chunks with z coords 1, 2, 5, and 6. The `w` component is time.
  - `maxPrefetchChunks` is the max number of total prefetch requests that can be made after any load. The loader will always stop issuing prefetch requests at this number even if `maxPrefetchDistance` would allow loading more.

### Other changes:

- OmeZarrLoader.ts was getting long and cumbersome, and began with a growing string of items that made the actual beginning of the `OmeZarrLoader` class declaration hard to find. I moved some of these items into their own files in a new /zarr_utils subdirectory:
  - WrappedStore.ts for the object that lets us link into zarrita's low-level chunk request behavior.
  - types.ts for stray shared types and the big declaration of the OME-Zarr JSON metadata format.
- /zarr_utils also has one brand new item: `ChunkPrefetchIterator` encapsulates the somewhat complicated process of identifying which chunks can and should be prefetched and in what order.
- Did a bit of refactoring in `OmeZarrLoader` to make all of the above easier, notably adding a couple methods for convenient reshuffling of dimension-ordered arrays.

### CODE REVIEW

File: src/utils/RequestQueue.ts [hunk 1]
```diff
@@ -30,13 +30,24 @@ interface RequestItem<V> {
  * while the original request is still in the queue.
  */
 export default class RequestQueue {
-  /** The maximum number of requests that can be handled concurently.
-    Once reached, additional requests will be queued up to run once a running request completes.*/
+  /**
+   * The maximum number of requests that can be handled concurently.
```
Comment (toloudis) [inline src/utils/RequestQueue.ts:34]:
```suggestion
   * The maximum number of requests that can be handled concurrently.
```

File: src/utils/RequestQueue.ts [hunk 2]
```diff
@@ -30,13 +30,24 @@ interface RequestItem<V> {
...
+   * Once reached, additional requests will be queued up to run once a running request completes.
+   */
   private maxActiveRequests: number;
 
+  /**
+   * The maximum number of requests that can be handled concurrently for low-priority requests to be queued. Equal to
+   * `maxActiveRequests` by default, but may be set lower to ensure the queue leaves space for high-priority requests.
```
Comment (toloudis) [inline src/utils/RequestQueue.ts:41]: seems like the default should always be to allow high priority requests and keep this one more conservative?

File: src/utils/RequestQueue.ts [hunk 3]
```diff
@@ -119,23 +140,31 @@ export default class RequestQueue {
...
+        this.addRequestToQueue(key, lowPriority);
+      }
+    } else {
+      const lowPriorityIndex = this.queueLowPriority.indexOf(key);
+      if (lowPriorityIndex > -1 && !lowPriority) {
+        // This request is registered and queued, but is now being requested with high priority.
+        // Promote it to high priority.
+        this.queueLowPriority.splice(lowPriorityIndex, 1);
         this.addRequestToQueue(key);
+      } else if (delayMs <= 0) {
```
Comment (frasercl) [inline src/utils/RequestQueue.ts:163]: Might be worth thinking more deeply about how "promote to high priority" and "cancel delay and execute immediately" might interact here.

File: src/loaders/OmeZarrLoader.ts [hunk 4]
```diff
@@ -424,10 +354,63 @@ class OMEZarrLoader implements IVolumeLoader {
...
+    // Get number of chunks per dimension in `scaleLevel`
+    const chunkDimsUnordered = scaleLevel.shape.map((dim, idx) => Math.ceil(dim / scaleLevel.chunks[idx]));
+    const chunkDims = this.orderByTCZYX(chunkDimsUnordered, 1);
+
+    const subscriber = this.requestQueue.addSubscriber();
+    // `ChunkPrefetchIterator` yields chunk coordinates in order of roughly how likely they are to be loaded next
+    const prefetchIterator = new ChunkPrefetchIterator(
+      chunkCoords,
+      this.fetchOptions.maxPrefetchDistance,
+      new Vector4(chunkDims[4], chunkDims[3], chunkDims[2], chunkDims[0])
```
Comment (frasercl) [inline src/loaders/OmeZarrLoader.ts:394]: Consider: should we use some other type than `Vector4`?

File: src/loaders/zarr_utils/ChunkPrefetchIterator.ts [hunk 5]
```diff
@@ -0,0 +1,104 @@
...
+        }
+      }
+    }
+
+    // Create `PrefetchDirectionState`s for each direction and fill them with chunks at the borders of the fetched set
+    const directions: PrefetchDirectionState[] = extrema.map((start, direction) => {
+      const end =
+        direction & 1
+          ? Math.min(start + maxOffsetArr[direction >> 1], chunkSizeArr[direction >> 1] - 1)
+          : Math.max(start - maxOffsetArr[direction >> 1], 0);
```
Comment (frasercl) [inline src/loaders/zarr_utils/ChunkPrefetchIterator.ts:64]: clean up and/or document this! it's dense and hard to read this way

File: src/loaders/zarr_utils/ChunkPrefetchIterator.ts
Comment (frasercl) [inline src/loaders/zarr_utils/ChunkPrefetchIterator.ts:1]: unit test this!

File: src/loaders/zarr_utils/types.ts [hunk 6]
```diff
@@ -0,0 +1,69 @@
...
+export type TCZYX<T> = [T, T, T, T, T];
+export type SubscriberId = ReturnType<SubscribableRequestQueue["addSubscriber"]>;
+
+export type OMECoordinateTransformation =
+  | {
+      type: "identity";
+    }
+  | {
+      type: "translation";
+      translation: number[];
```
Comment (ShrimpCryptid) [inline src/loaders/zarr_utils/types.ts:12]: Do these arrays have specific lengths? (ZYX?) Not sure if it's useful to capture in the type, just thinking.

File: src/utils/SubscribableRequestQueue.ts [hunk 7]
```diff
@@ -27,8 +27,14 @@ export default class SubscribableRequestQueue {
...
-  constructor(maxActiveRequests?: number) {
-    this.queue = new RequestQueue(maxActiveRequests);
+  constructor(maxActiveRequests?: number, maxLowPriorityRequests?: number);
+  constructor(inner: RequestQueue);
+  constructor(maxActiveRequests?: number | RequestQueue, maxLowPriorityRequests?: number) {
+    if (typeof maxActiveRequests === "number" || maxActiveRequests === undefined) {
+      this.queue = new RequestQueue(maxActiveRequests, maxLowPriorityRequests);
+    } else {
+      this.queue = maxActiveRequests;
+    }
```
Comment (ShrimpCryptid) [inline src/utils/SubscribableRequestQueue.ts:37]: First time I've seen an overload like this, interesting. Maybe useful to add a docstring comment to explain that it can take in an existing RequestQueue as a parameter

File: src/loaders/zarr_utils/types.ts (hunk 6 above)
Comment (frasercl) [reply src/loaders/zarr_utils/types.ts:12]: OME-Zarr technically allows any number of dimensions in any order. (Many of the hoops this loader has to jump through have to do with converting arbitrary dimension orderings from metadata into something consistent - see `orderByDimension` and `orderByTCZYX` below. I mostly try to type that converted, consistent representation as a fixed-length array.)

File: src/test/RequestQueue.test.ts [hunk 8]
```diff
@@ -218,12 +218,79 @@ describe("test RequestQueue", () => {
         { key: "b", requestAction: work },
       ];
       const start = Date.now();
-      const promises = rq.addRequests(requests, delayMs);
+      const promises = rq.addRequests(requests, false, delayMs);
```
Comment (ShrimpCryptid) [inline src/test/RequestQueue.test.ts:221]: Nit: Seeing this makes me think it would be nice to have an enum for the request type (HIGH or LOW), since there's no context for it when seeing it without Intellisense... Very likely unnecessarily verbose though.

File: src/test/RequestQueue.test.ts [hunk 9]
```diff
@@ -218,12 +218,79 @@ describe("test RequestQueue", () => {
...
+      const prom1 = rq.addRequest("a", work);
+      const prom2 = rq.addRequest("b", work, true);
+      expect(rq.requestRunning("a")).to.be.true;
+      expect(rq.requestRunning("b")).to.be.false;
+      rq.addRequest("b", work);
+      expect(rq.requestRunning("b")).to.be.true;
+      await prom1;
+      await prom2;
+      expect(count).to.equal(2);
+    });
```
Comment (ShrimpCryptid) [inline src/test/RequestQueue.test.ts:293]: I know the current behavior for RequestQueue requires all high priority tasks to be complete before moving onto the low priority tasks... Would this ever cause problematic stalling?

Like, say a high priority task is stuck due to a network timeout or something. This would prevent all low priority tasks for being executed, even if there is capacity available to handle them.

Comment (ShrimpCryptid) [reply src/test/RequestQueue.test.ts:293]: Okay, correction. Took a closer look at RequestQueue, and it allows low priority requests to be started once the total number of active requests dips below the max low priority count and if the high priority queue is empty. Question answered 👍

File: src/loaders/zarr_utils/ChunkPrefetchIterator.ts [hunk 10]
```diff
@@ -0,0 +1,119 @@
...
+  directions: PrefetchDirectionState[];
+
+  constructor(chunks: TCZYX<number>[], tzyxMaxOffset: TZYX, tzyxChunkSize: TZYX) {
+    // Get max and min chunk coordinates for T/Z/Y/X
+    const extrema = [Infinity, -Infinity, Infinity, -Infinity, Infinity, -Infinity, Infinity, -Infinity];
+
+    for (const chunk of chunks) {
+      for (let j = 0; j < 4; j++) {
+        const val = chunk[skipC(j)];
+        const extremaIdx = j << 1;
```
Comment (ShrimpCryptid) [inline src/loaders/zarr_utils/ChunkPrefetchIterator.ts:54]: Nit: More readable to as `j * 2`?

File: src/loaders/OmeZarrLoader.ts [hunk 11]
```diff
@@ -215,44 +103,62 @@ function convertChannel(channelData: zarr.TypedArray<zarr.NumberDataType>): Uint
...
+  /** The max. number of total chunks that can be prefetched after any load. */
+  maxPrefetchChunks: number;
+};
 
-class OMEZarrLoader implements IVolumeLoader {
-  scaleLevels: NumericZarrArray[];
-  multiscaleMetadata: OMEMultiscale;
-  omeroMetadata: OmeroTransitionalMetadata;
-  axesTCZYX: [number, number, number, number, number];
+const DEFAULT_FETCH_OPTIONS = {
```
Comment (ShrimpCryptid) [inline src/loaders/OmeZarrLoader.ts:126]:
```suggestion
const DEFAULT_FETCH_OPTIONS: ZarrLoaderFetchOptions = {
```

**Review by ShrimpCryptid — APPROVED** [review]
LGTM!

File: src/loaders/zarr_utils/types.ts [hunk 12]
```diff
@@ -0,0 +1,69 @@
...
+  unit?: string;
+};
+
+export type OMEDataset = {
+  path: string;
+  coordinateTransformations?: OMECoordinateTransformation[];
+};
+
+// https://ngff.openmicroscopy.org/latest/#multiscale-md
+export type OMEMultiscale = {
```
Comment (toloudis) [inline src/loaders/zarr_utils/types.ts:35]: Great job on making these types.  Kinda surprising that there isn't a thirdparty js lib that captures all these :). Also it's unfortunate that the ome spec is a good deal more general than the sorts of data we hope to present in this viewer. (example - axis names can be just about anything)

File: src/loaders/zarr_utils/types.ts [hunk 13]
```diff
@@ -0,0 +1,69 @@
...
+  id: number;
+  name: string;
+  version: string;
+  channels: {
+    active: boolean;
+    coefficient: number;
+    color: string;
+    family: string;
+    inverted: boolean;
+    label: string;
```
Comment (toloudis) [inline src/loaders/zarr_utils/types.ts:56]: Sadly, this "transitional" data structure is also the only official vehicle for getting channel names in.

File: src/test/ChunkPrefetchIterator.test.ts [hunk 14]
```diff
@@ -0,0 +1,164 @@
+import { expect } from "chai";
+
+import { TCZYX } from "../loaders/zarr_utils/types";
+import ChunkPrefetchIterator from "../loaders/zarr_utils/ChunkPrefetchIterator";
+
+const EXPECTED_3X3X3X3 = [
```
Comment (toloudis) [inline src/test/ChunkPrefetchIterator.test.ts:6]: it's hard to understand why this is called 3x3x3 when it has all of t,z,y,x and the arrays are 5 elements each.

File: src/loaders/zarr_utils/ChunkPrefetchIterator.ts (hunk 10 above)
Comment (toloudis) [reply src/loaders/zarr_utils/ChunkPrefetchIterator.ts:54]: agree

Comment (toloudis) [reply src/loaders/zarr_utils/ChunkPrefetchIterator.ts:54]: Nit: I get what this is doing, but this code is a bit hard to parse when reading it.   I feel like it might be even more readable if the loop were unrolled and/or the extrema array were more structured, since the dims are always t,z,y,x.   It's a 4D bounding box!  The skipC function is hard to understand too.
pseudocode : 
for each chunk:  
   update min/max of T value
   update min/max of Z value
   update min/max of Y value
   update min/max of X value

File: src/loaders/zarr_utils/ChunkPrefetchIterator.ts [hunk 15]
```diff
@@ -0,0 +1,119 @@
...
+/**
+ * Since the user is most likely to want nearby data (in space or time) first, we should prefetch those chunks first.
+ *
+ * Given a list of just-loaded chunks and some bounds, `ChunkPrefetchIterator` iterates evenly outwards in T/Z/Y/X.
+ */
+// NOTE: Assumes `chunks` form a rectangular prism! Will create gaps otherwise! (in practice they always should)
+export default class ChunkPrefetchIterator {
+  directions: PrefetchDirectionState[];
+
+  constructor(chunks: TCZYX<number>[], tzyxMaxOffset: TZYX, tzyxChunkSize: TZYX) {
```
Comment (toloudis) [inline src/loaders/zarr_utils/ChunkPrefetchIterator.ts:47]: is tzyxMaxOffset just the highest index for chunks in those dims?

File: src/loaders/zarr_utils/ChunkPrefetchIterator.ts [hunk 16]
```diff
@@ -0,0 +1,119 @@
...
+      this.directions = this.directions.filter((dir) => {
+        if (dir.direction & 1) {
+          return dir.start + offset <= dir.end;
+        } else {
+          return dir.start - offset >= dir.end;
+        }
+      });
+
+      // Yield chunks one chunk farther out in every remaining direction
+      for (const dir of this.directions) {
```
Comment (toloudis) [inline src/loaders/zarr_utils/ChunkPrefetchIterator.ts:108]: to make sure I understand, is there currently an implied preferential order in T,Z,Y,X ?

File: src/loaders/zarr_utils/ChunkPrefetchIterator.ts (hunk 15 above)
Comment (frasercl) [reply src/loaders/zarr_utils/ChunkPrefetchIterator.ts:47]: `tzyxChunkSize` is what you're describing; `tzyxMaxOffset` is the maximum number of chunks to move outward in both directions, so e.g. if you're at T = 30 and `tzyxMaxOffset` in T is 2, we prefetch T = 28, 29, 31, 32. (These are both bad names and I'm open to suggestions. And they ought to be documented better regardless!)

File: src/loaders/zarr_utils/ChunkPrefetchIterator.ts (hunk 16 above)
Comment (frasercl) [reply src/loaders/zarr_utils/ChunkPrefetchIterator.ts:108]: Yes

File: src/loaders/zarr_utils/ChunkPrefetchIterator.ts [hunk 17]
```diff
@@ -0,0 +1,119 @@
...
+        }
+
+        if (val > extrema[extremaIdx + 1]) {
+          extrema[extremaIdx + 1] = val;
+        }
+      }
+    }
+
+    // Create `PrefetchDirectionState`s for each direction
+    const directions: PrefetchDirectionState[] = extrema.map((start, direction) => {
```
Comment (toloudis) [inline src/loaders/zarr_utils/ChunkPrefetchIterator.ts:67]: Just writing this to help myself understand.  So the directions are either toward the max or toward the min for each of 4 dims. In my head I am turning the bit operations into little functions.  (`isPositive(direction)`, `getAxisDimension(direction)` )

File: src/loaders/zarr_utils/ChunkPrefetchIterator.ts (hunk 15 above)
Comment (toloudis) [reply src/loaders/zarr_utils/ChunkPrefetchIterator.ts:47]: I think something like `tzyxMaxChunkIndexOffset` is slightly more explicit

Comment (toloudis) [reply src/loaders/zarr_utils/ChunkPrefetchIterator.ts:47]: or tzyxPrefetchOffsets

File: src/loaders/zarr_utils/ChunkPrefetchIterator.ts (hunk 17 above)
Comment (ShrimpCryptid) [reply src/loaders/zarr_utils/ChunkPrefetchIterator.ts:67]: Now that you've said it I'd love for these to be functions instead of bit operations 🥺

File: src/loaders/OmeZarrLoader.ts [hunk 18]
```diff
@@ -215,44 +103,62 @@ function convertChannel(channelData: zarr.TypedArray<zarr.NumberDataType>): Uint
...
+  maxPrefetchDistance: [5, 5, 5, 5] as [number, number, number, number],
+  maxPrefetchChunks: 30,
```
Comment (toloudis) [inline src/loaders/OmeZarrLoader.ts:128]: Idea/ Future suggestion: make this match the defaults explicitly passed in by the test app, but allow them to be tweaked at runtime using the advanced settings gui

**Review by toloudis — APPROVED** [review]
No show stoppers for this PR, just some suggestions for clarification/readability.
